index.bs

```diff
@@ -205,9 +207,22 @@ ISSUE: Clarify the behavior of inline sessions
 
 Automatic composition {#automatic-composition}
 ---------
-<dfn>performs automatic composition</dfn>
 
-ISSUE: Add explanation here
+An [=xr device=] during an {{XRSessionMode/"immersive-ar"}} session will have its [=xr compositor=] <dfn>perform automatic composition</dfn> by drawing the rendered scene on top of the real-world background.
+
+In case the [=xr device=] is using a [=passthrough camera=], it MUST alpha-blend non-opaque rendered pixels on top of the [=passthrough camera=]'s [=aligned video stream=]. Such a device MUST have an {{XRSession/environmentBlendMode}} of {{alpha-blend}}.
```

Comment:
What does a device that optionally supports video passthrough (like a Vive Pro or Varjo XR-1) return for `XRSession.environmentBlendMode`?
1. Do we return `"opaque"` if you created an `"immersive-vr"` session (which _does not_ turn on the passthrough camera) and `"alpha-blend"` if you created an `"immersive-ar"` session (which _does_ turn on the passthrough camera)?
1. Do we return `"alpha-blend"` even for `"immersive-vr"` sessions, because you _could have_ turned on video passthrough by creating an `"immersive-ar"` session?
1. Do we return `"opaque"` even for `"immersive-ar"` sessions, because it is the "natural" state of the device?

Once a developer is already in an `XRSession`, I believe they will care most about how that session will actually behave.  My suggestion is that `XRSession.environmentBlendMode` be defined to represent the **"effective environment blend mode"** for the active session, informed both by the XR device that's in use and the `XRSessionMode` that the app selected.

This is halfway stated in the bullets above:
> The {{XRSession/environmentBlendMode}} MUST NOT be {{opaque}} for {{XRSessionMode/"immersive-ar"}} sessions.

> When an [=xr device=] [=performs automatic composition=] using a [=passthrough camera=] in an {{XRSessionMode/"immersive-ar"}} session, it MUST report {{XREnvironmentBlendMode/"alpha-blend"}} blending behavior.

We may need to explicitly state the opposite in the `"opaque"` bullet above, such as:

> When an [=xr device=] with a [=passthrough camera=] that's in an {{XRSessionMode/"immersive-ar"}} session is not [=performs automatic composition|performing automatic composition=], it MUST report {{XREnvironmentBlendMode/"opaque"}} blending behavior.

And then here, we may want to be more explicit that the `environmentBlendMode` is only `"alpha-blend"` when in such an `"immersive-ar"` session:
> Such a device MUST have an {{XRSession/environmentBlendMode}} of {{alpha-blend}} **if and only if that session is an {{XRSessionMode/"immersive-ar"}} session**.

Reply:
IMO environmentBlendMode should be the effective environmentBlendMode, you're not querying it to see what capabilities the device has, you're querying it to see  how the device has chosen to behave in the current moment.


Why would an immersive-ar session turn off the passthrough camera by default?

Reply:
Is there a real case in which a device which supports alpha blend would not do it in immersive-ar?  Isn't that kinda the point of immersive-ar?

Reply:
> Is there a real case in which a device which supports alpha blend would not do it in immersive-ar? Isn't that kinda the point of immersive-ar?

I agree. If an author wants alpha-blend they should always select immersive-ar.
For that reason, we should not have `environmentBlendMode` in an immersive-vr session. It should always be `opaque`. 

For immersive-ar use cases, I can't think of a reason to have `opaque` as an option since such a session would not make sense.

So, let's remove `opaque` from the enum and add `environmentBlendMode` to a session only if it was created with `immersive-ar`.

Reply:
> So, let's remove opaque from the enum and add environmentBlendMode to a session only if it was created with immersive-ar.

Wanted to chime in and mention that I'm not aware of any other web APIs that function that way. If we add an attribute to `XRSession` it will be available to all `XRSession` instances. The only way you get around that is if we define an inherited type that's only returned for `immersive-ar`, and frankly that seems extreme simply to avoid exposing one enum.

> This avoids having 2 different ways for an author to generate AR content.

The point of moving the enum to the AR module was to avoid that concern all together. If it's introduced at the same time as the `immersive-ar` mode then there's no motivation to try and "fake" AR mode with `immersive-vr`. Plus, Magic Leap can further encourage the desired behavior here by simply not supporting `immersive-vr` sessions, which is the plan of record if I recall correctly, yes?

Reply:
> Wanted to chime in and mention that I'm not aware of any other web APIs that function that way. If we add an attribute to `XRSession` it will be available to all `XRSession` instances. The only way you get around that is if we define an inherited type that's only returned for `immersive-ar`, and frankly that seems extreme simply to avoid exposing one enum.

Yes, you're right.
So since we're likely going to add a number of extensions to XRSession/XRFrame, we should think about devising some way to pass that unknown information without having to extend the class. I don't really know what that would look like. (map<string, any>?)

For instance, both the plane and mesh proposals are extending XRFrame. I didn't think it through that this attribute would also be there if you didn't use those optional features.

> > This avoids having 2 different ways for an author to generate AR content.
> 
> The point of moving the enum to the AR module was to avoid that concern all together. If it's introduced at the same time as the `immersive-ar` mode then there's no motivation to try and "fake" AR mode with `immersive-vr`. 

What if a popular framework always asks for `immersive-vr` and starts keying off the blendmode? We can put wording in the spec but that won't stop them if they think it's easier (or they just made a mistake and then never bother to fix it)

> Plus, Magic Leap can further encourage the desired behavior here by simply not supporting `immersive-vr` sessions, which is the plan of record if I recall correctly, yes?

We don't right now but if a good portion of AR content takes that other path, we would eventually be forced to support it (unless we take over the entire AR market :-P)

Reply:
#15

Reply:
> I didn't think it through that this attribute would also be there if you didn't use those optional features.

We'd have them be null in VR mode. We could let the EBM be null, but that's not functionally different from having an "opaque" variant, and typically it's better to have an enum variant for the null case instead of making an enum nullable.

Reply:
> > I didn't think it through that this attribute would also be there if you didn't use those optional features.
> 
> We'd have them be null in VR mode. We could let the EBM be null, but that's not functionally different from having an "opaque" variant, and typically it's better to have an enum variant for the null case instead of making an enum nullable.

Yes and I found other places in the web platform that includes optional attributes, so I _think_ we can make it work:

```
partial interface XRSession {
  readonly attribute XREnvironmentBlendMode? environmentBlendMode;
}
```
note the '?'
If there is no reason for authors to check it during a VR session, we should not provide.

Reply:
> Yes and I found other places in the web platform that includes optional attributes, so I think we can make it work:

Right, it's totally _possible_ to do this, but optional _enums_ are rare because it's more descriptive to have the null variant be an actual variant.

Reply:
> > Yes and I found other places in the web platform that includes optional attributes, so I think we can make it work:
> 
> Right, it's totally _possible_ to do this, but optional _enums_ are rare because it's more descriptive to have the null variant be an actual variant.

OK. Maybe we should list it as another issue, we'll present the options with their pros/cons and put it to a vote.

Reply:
> > Plus, Magic Leap can further encourage the desired behavior here by simply not supporting `immersive-vr` sessions, which is the plan of record if I recall correctly, yes?
> 
> We don't right now but if a good portion of AR content takes that other path, we would eventually be forced to support it (unless we take over the entire AR market :-P)

What would lead you to believe that a good portion of AR content WOULD take that other path ("asking for immersive-vr and being fine with a non-opaque blend mode" is the other path you're referring to, right?) ?  Such a path would deliberately put them in a non-AR mode in a handheld device, for example.  Would you be more comfortable having this be an option that has to be explicit?

I *do* think that you're disregarding the case where a developer just wants to present a model, and is less worried about whether that's AR in a see-thru headset or VR in a traditional VR headset; for example, if I write an app to render an exploded model of an engine, I'd be fine with it being presented in see-thru AR if that's how you can present it, or I'd be fine with it in VR if that's the mode available, too.  We shouldn't make this impossible, or even hard; we SHOULD make it clear to the developer what they're asking for, since that would change what a hand-held device would do.  Although I understand the desire from your perspective to not simply support immersive-vr, I do expect some other see-thru headset vendor might believe that supporting immersive-vr is the right choice for them to make.

Reply:
> > > Yes and I found other places in the web platform that includes optional attributes, so I think we can make it work:
> > 
> > 
> > Right, it's totally _possible_ to do this, but optional _enums_ are rare because it's more descriptive to have the null variant be an actual variant.
> 
> OK. Maybe we should list it as another issue, we'll present the options with their pros/cons and put it to a vote. I'm happy to help out with that.

As this is the initial module PR, I think we should go with this, and yes, file an issue:  "Clarify developer AR/VR intent paths" with choices about the modes, what they might end up in on different devices, and what use cases they're for, and let's discuss in greater detail.

Reply:
> > We don't right now but if a good portion of AR content takes that other path, we would eventually be forced to support it (unless we take over the entire AR market :-P)
> 
> What would lead you to believe that a good portion of AR content WOULD take that other path 

Authors will take the easiest way. If the AR device they own returns `additive` as a blend mode, they will use that and ignoring asking for 'immersive-ar'.
@toji 's explainer had this as an example so they might think it's ok to do.

(To reiterate: my problem with keying of the blendmode attribute is that this allows 2 different workflows to create AR content)

>("asking for immersive-vr and being fine with a non-opaque blend mode" is the other path you're referring to, right?) ? 

yes

> Such a path would deliberately put them in a non-AR mode in a handheld device, for example. Would you be more comfortable having this be an option that has to be explicit?

Yes. If there was a way for an author to say "I know what I'm doing and will request AR AND VR" in 1 call that would be great.
That does bring up the question what a device that can do both would do. Would we let that decision be UA dependent?

> I _do_ think that you're disregarding the case where a developer just wants to present a model, and is less worried about whether that's AR in a see-thru headset or VR in a traditional VR headset; for example, if I write an app to render an exploded model of an engine, I'd be fine with it being presented in see-thru AR if that's how you can present it, or I'd be fine with it in VR if that's the mode available, too. We shouldn't make this impossible, or even hard; we SHOULD make it clear to the developer what they're asking for, since that would change what a hand-held device would do. 

The current API allows that; the author just has to ask for AR first and if not supported, then for VR. (I'm assuming that by "I" in this paragraph, you mean "the author" and not "the user".)

It's a bit clunky but a reasonable decision.

> Although I understand the desire from your perspective to not simply support immersive-vr, I do expect some other see-thru headset vendor might believe that supporting immersive-vr is the right choice for them to make.

As long as they don't return `additive` for blend mode, that would be a reasonable choice.

As for handheld devices, is Chrome going to allow `immersive-vr` even on devices that don't support VR?

Reply:
This PR has been merged and further conversation here is entirely likely to be missed by the relevant parties.  I have moved this this topic to its own dedicated issue. https://github.com/immersive-web/webxr-ar-module/issues/16
Please continue discussion there.

Reply:
> As for handheld devices, is Chrome going to allow immersive-vr even on devices that don't support VR?

The answer depends on what you mean by "devices that don't support VR"? While we will support Daydream on devices that are Daydream Ready, our aim is to support `immersive-vr` on as many other mobile devices as possible via Cardboard mode.

We have _no plans_ at this time to have a "fullscreen VR" mode (ie: mobile AR style presentation without the passthrough feed.) Nor do we intend to support `immersive-vr` on desktops/laptops that have no VR hardware attached.

Reply:
@toji, please move this comment to issue #16.